src/System.Slices/src/System/Contract.cs

```diff
@@ -31,17 +31,26 @@ public static void Requires(bool condition)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void RequiresInRange(int start, int length)
         {
-            if (!(start >= 0 && start < length))
+            // explicit unchecked block is used in order to avoid OverflowException when casting 
+            // negative int to unsigned int when the code is compiled with /checked option
+            unchecked
             {
-                throw new ArgumentOutOfRangeException();
+                // use unsigned int to reduce range check by one ( >= 0 )
+                if ((uint)start >= (uint)length)
```

Comment:
I'm not the final call here - I'm just helping highlight potential issues :smile: 

Thanks for doing the homework - and I agree, every cycle matters on the hot path.

Can we get unit tests to prove this is the correct decision?